src/client/Microsoft.Identity.Client/ManagedIdentity/CsrRequest.cs

```diff
@@ -0,0 +1,439 @@
+        /// <param name="tenantId">AAD tenant_id.</param>
+        /// <param name="cuid">CuidInfo object containing VMID and VMSSID.</param>
+        /// <returns>CsrRequest containing the PEM CSR.</returns>
+        public static CsrRequest Generate(string clientId, string tenantId, CuidInfo cuid)
+        {
+            if (string.IsNullOrWhiteSpace(clientId))
+                throw new ArgumentException("clientId must not be null or empty.", nameof(clientId));
+            if (string.IsNullOrWhiteSpace(tenantId))
+                throw new ArgumentException("tenantId must not be null or empty.", nameof(tenantId));
+            if (cuid == null)
+                throw new ArgumentNullException(nameof(cuid));
+            if (string.IsNullOrWhiteSpace(cuid.Vmid))
+                throw new ArgumentException("cuid.Vmid must not be null or empty.", nameof(cuid.Vmid));
+            if (string.IsNullOrWhiteSpace(cuid.Vmssid))
+                throw new ArgumentException("cuid.Vmssid must not be null or empty.", nameof(cuid.Vmssid));
+
+            string pemCsr = GeneratePkcs10Csr(clientId, tenantId, cuid);
+            return new CsrRequest(pemCsr);
+        }
+
+        /// <summary>
+        /// Generates a PKCS#10 Certificate Signing Request in PEM format.
+        /// </summary>
+        private static string GeneratePkcs10Csr(string clientId, string tenantId, CuidInfo cuid)
+        {
+            // Generate RSA key pair (2048-bit)
+            RSA rsa = CreateRsaKeyPair();
+
+            try
+            {
+                // Build the CSR components
+                byte[] certificationRequestInfo = BuildCertificationRequestInfo(clientId, tenantId, cuid, rsa);
+                byte[] signatureAlgorithm = BuildSignatureAlgorithmIdentifier();
+                byte[] signature = SignCertificationRequestInfo(certificationRequestInfo, rsa);
+
+                // Combine into final CSR structure
+                byte[] csrBytes = BuildFinalCsr(certificationRequestInfo, signatureAlgorithm, signature);
+
+                // Convert to PEM format
+                return ConvertToPem(csrBytes);
+            }
+            finally
+            {
+                rsa?.Dispose();
+            }
+        }
+
+        /// <summary>
+        /// Creates a 2048-bit RSA key pair compatible with all target frameworks.
+        /// </summary>
+        private static RSA CreateRsaKeyPair()
+        {
+#if NET462 || NET472
+            var rsa = new RSACryptoServiceProvider(2048);
+            return rsa;
+#else
+            var rsa = RSA.Create();
+            rsa.KeySize = 2048;
+            return rsa;
+#endif
+        }
+
+        /// <summary>
+        /// Builds the CertificationRequestInfo structure containing subject, public key, and attributes.
+        /// </summary>
+        private static byte[] BuildCertificationRequestInfo(string clientId, string tenantId, CuidInfo cuid, RSA rsa)
```

Comment:
We should use CertificateRequest where available, and a thin polyfill for old TFMs like net462 and netstd2

Reply:
We can't use a polyfil
 
System.Security.Cryptography.X509Certificates.CertificateRequest` cannot be used because:
 
- It only supports X.509 extensions, not PKCS#10 attributes
- The CUID data must be embedded as an attribute (OID 1.2.840.113549.1.9.7) in the CSR attributes section
- This is a fundamental limitation that requires custom ASN.1 encoding

Reply:
https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.certificaterequest.otherrequestattributes?view=net-9.0#system-security-cryptography-x509certificates-certificaterequest-otherrequestattributes

Reply:
Resolved after discussion with Managed Crypto Discussion. Current approach is fine.